### ISSUE TITLE
Possible bug in default implementation of 'mean_squared_error' (MSE) loss function in Keras.

### ISSUE THREAD

**Issue by wzywiec** [rstudio/keras#1071]
In the Boston Housing Prices regression tutorial, I tested the default model and noted an apparent source of instability in the model.

I looked into it, and figured out that the last batch only has 3 samples. When I started tracking the loss of each batch, I noted that the loss of the last batch was a factor of approximately 11 greater than the losses of all other batches (plot below).

This makes sense, since 11 is roughly equal to the ratio of the default batch size to the last batch size (32/3 = 10.67). However, I've been poking around in the keras/tf backend and it's unclear whether or not the loss of the last batch is ever adjusted/scaled prior to computing the gradients.

As an empirical test, I built a sum of squared errors loss function (MSE except I skipped the last step of calculating the mean) and applied it to the dataset. The result was apparent and significant improvement in the model, which leads me to believe that the unadjusted last batch loss is causing the model to overfit and become unstable (plots also below).

Basically, I'm posing 2 questions:
1) Does the default implementation of MSE scale/adjust the loss of the last batch prior to computing the gradients?
2) Either way, can someone point me to the exact place(s) in the code where the loss is used to compute the gradient?

(These questions also equally apply to MAE)

Thanks!

![image](https://user-images.githubusercontent.com/30445407/85391271-cb140080-b4fe-11ea-8212-9532646fd0cc.png)

![image](https://user-images.githubusercontent.com/30445407/85390187-4c6a9380-b4fd-11ea-8fe6-d57b351cdee2.png)

**Comment by dfalbel** [comment]
AFAIK keras will compute the mean of the gradients for each observation in the minibatch. 
Also, Keras shuffles observations between epochs by default, and it cause an even larger variance in the last minibatch loss.

The exact definition of MSE can be found here: 

https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/losses.py#L1169-L1195

**Comment by wzywiec** [comment]
Your first statement that Keras computes the mean of the gradients is interesting. I'll check out the TensorBoard and keep looking deeper into TensorFlow.

I did turn shuffling on and off to test some other ideas I had (plots below).

I would expect to see larger variance in the last batch, because it's smaller, but with MSE plots in my first post, there is a much higher loss in batch 11, which I think needs to be adjusted/scaled prior to computing the gradients.

When I used the sum of squared errors (SSE) loss, the average loss of batch 11 is the same as batches 1-10, and more importantly, the resultant MAE of the training and validation data consistently performs better and is less noisy.

This seems to indicate that the loss of the last batch is not adjusted/scaled. I thought this might just be a Keras/TF issue with R, but I also see the same thing in the example in Deep Learning with Python.

If I find the answer in the meantime, I'll post it here.

![image](https://user-images.githubusercontent.com/30445407/85482103-d6574280-b577-11ea-8f34-be3ae98be89b.png)

![image](https://user-images.githubusercontent.com/30445407/85482927-bb390280-b578-11ea-9dbc-6cb43f761ff9.png)

**Comment by dfalbel** [comment]
Here's a small example where you can see exactly hat happens:

```
library(keras)

x <- matrix(1:5, ncol = 1)
y <- c(1:4, 6)

model <- keras_model_sequential() %>% 
  layer_dense(units = 1, use_bias = FALSE, kernel_initializer = initializer_constant(1))

model %>% compile(loss = "mse", optimizer = optimizer_sgd(lr = 1))

model %>% fit(x, y, batch_size = 3, epochs = 1, shuffle = FALSE)

get_weights(model)
```

After the first epoch, the `weight` value is 6 which can be computed exactly by making the mean of the gradients of the last 2 observations. Ie.

```
w <- 1
w_new = w - 1/2*(-2*x[4]*(y[4] - w*x[4]) - 2*x[5]*(y[5] - w*x[5]))
```

**Comment by wzywiec** [comment]
I tracked the biases and weights in Keras using callbacks (code below) and then manually performed a forward pass and backpropagation to compute the gradients and bias/weight updates.

I confirmed that the software computes the gradient of each sample, and then divides each batch update by the batch size, which means the bias/sensitivity is left unadjusted. In other words, if you have a smaller last batch, the average "per sample weight" of the last batch can potentially be a couple order of magnitude greater than all other batches, which causes the model to overfit.

I'm going to close out the issue, because my questions were answered, but the results and math appear to indicate that MSE isn't as good as SSE for neural network regression problems, specifically in cases where the number of training samples is not evenly divisible by batch size.

Thanks, Daniel for responding so quickly!

LossHistory <- R6::R6Class('LossHistory',
  inherit = KerasCallback,
  public = list(
    loss = NULL,
    on_batch_end = function(batch, logs = list()) {
      self$loss<- c(self$loss, logs[["loss"]])
    }))

WeightHistory <- R6::R6Class('WeightHistory',
  inherit = KerasCallback,
  public = list(
    weight = list(),
    on_batch_end = function(batch, logs = list()) {
      self$weight <- c(self$weight, get_weights(model))
    }))

loss.history <- LossHistory$new()
weight.history <- WeightHistory$new()

history <- model %>% fit(
  train_data,
  train_labels,
  epochs = epochs,
  shuffle = FALSE,
  validation_split = 0.2,
  verbose = 0,
  callbacks = list(loss.history, weight.history)
)
